core/state_processor.go

```diff
@@ -84,8 +84,16 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 		allLogs = append(allLogs, receipt.Logs...)
 	}
 	// Finalize the block, applying any consensus engine specific extras (e.g. block rewards)
+	statedb.Prepare(common.Hash{}, block.Hash(), len(block.Transactions()))
 	p.engine.Finalize(p.bc, header, statedb, block.Transactions(), block.Uncles(), receipts, block.Randomness())
 
+	if len(statedb.GetLogs(common.Hash{})) > 0 {
```

Comment:
Not 100% clear on what's going on here. Would there be a single receipt for all blocks? If not, won't we have the null hash being mapped to many receipts?

Reply:
Yeah, many receipts will be associated into null hash, but they can be separated because they are contained in different blocks.

Comparison between using null and block hash
* if block hash is used, there will be more hashes without corresponding transactions
* if null hash is used, (tx hash, log index) pairs will no longer be unique

I'll next check that tx hash in receipt is not consensus data.